src/nvim/tui/tui.c

```diff
@@ -809,24 +809,27 @@ static void fix_terminfo(TUIData *data)
 #define TMUX_WRAP(seq) (inside_tmux ? "\x1bPtmux;\x1b" seq "\x1b\\" : seq)
   // Support changing cursor shape on some popular terminals.
   const char *term_prog = os_getenv("TERM_PROGRAM");
+  const char *vte_version = os_getenv("VTE_VERSION");
 
   if ((term_prog && !strcmp(term_prog, "iTerm.app"))
-      || os_getenv("ITERM_SESSION_ID") != NULL) {
-    // iterm
-    data->unibi_ext.enter_insert_mode = (int)unibi_add_ext_str(ut, NULL,
-        TMUX_WRAP("\x1b]50;CursorShape=1;BlinkingCursorEnabled=1\x07"));
-    data->unibi_ext.enter_replace_mode = (int)unibi_add_ext_str(ut, NULL,
-        TMUX_WRAP("\x1b]50;CursorShape=2;BlinkingCursorEnabled=1\x07"));
-    data->unibi_ext.exit_insert_mode = (int)unibi_add_ext_str(ut, NULL,
-        TMUX_WRAP("\x1b]50;CursorShape=0;BlinkingCursorEnabled=0\x07"));
-  } else {
-    // xterm-like sequences for blinking bar and solid block
+      || os_getenv("ITERM_SESSION_ID") != NULL
+      || os_getenv("XTERM_VERSION") != NULL
+      || (vte_version && atoi(vte_version) >= 3900)) {
+    // iTerm, xterm, and VTE based terminals support DESCCUSR.
     data->unibi_ext.enter_insert_mode = (int)unibi_add_ext_str(ut, NULL,
         TMUX_WRAP("\x1b[5 q"));
     data->unibi_ext.enter_replace_mode = (int)unibi_add_ext_str(ut, NULL,
         TMUX_WRAP("\x1b[3 q"));
     data->unibi_ext.exit_insert_mode = (int)unibi_add_ext_str(ut, NULL,
         TMUX_WRAP("\x1b[2 q"));
+  } else if ((term_prog && !strcmp(term_prog, "Konsole"))
+      || os_getenv("KONSOLE_DBUS_SESSION")) {
```

Comment:
Users can set sshd/tmux to inherit the TERM_PROGRAM env var for convenience.
I'm not sure why both checks are needed for iTerm, but for Konsole, TERM_PROGRAM isn't automatically set. Users can configure Konsole to set this if they want to.